Add support for Elli chargers

DerAndereAndi opened the pull request:
- Generic template schema fixes
- Added multiple workarounds for Elli charger specifics (define IP address in the configuration, use of `hasMeter` and `chargedEnergy` in the Elli templated marking lacking support of those)
- Add EEBUS template presets, one with meter and one without
- Move EEBUS `ski` definition to default params
- Add `ip` to default params
- Add 2 Elli templates, one for Connect (without meter) and one for Connect (with meter) and Pro
- Set the chargers voltage per charger. Currently hard coded, in the future this should be taken from the `site`
- Fixed excessive polling of external meters, will be fetched every 10s (could be improved by coupling this to the `site` `interval`)
- Fixed check for detecting if the EV is charging
- Fixed issue with self consumption handling
- Use evcc provided `machineid`/`plantid` when available
- Added patching of go library to accept buggy Elli EEBUS certificate

DerAndereAndi committed:
Template schema fixes

DerAndereAndi committed:
Workarounds for Elli chargers

- Add support for defining manual IP addresses for eebus chargers
- Add support for defining `hasMeter`
- Add support for defining `chargedEnergy`

DerAndereAndi committed:
Template updates

- Add eebus presets, one with meter and one without
- Move ski definition to default params
- Add ip default params
- Add 2 Elli templates, one for Connect (without meter) and one for Connect (with meter) and Pro

DerAndereAndi committed:
Update eebus repo reference

- Include updated zeroconf library
- Temporarily using the PRs original repo in mod

DerAndereAndi committed:
Set site voltage

Temporarily hardcode the site voltage. This should be taken from site in the future

DerAndereAndi committed:
Minor templates fix

DerAndereAndi committed:
Add fixed IP requirement

DerAndereAndi committed:
Add decorator for proper meter handling

StefanSchoof commented on templates/definition/defaults.yaml:
```diff
@@ -301,6 +314,30 @@ presets:
+      type: eebus
+      ski: {{ .ski }}
+      {{ if ne .ip "" }}ip: {{ .ip }}{{ end }}
+      Meter: false
```
Meter ist anders als die anderen Eigenschaften mit großem Anfangsbuchstaben. Ist das Absicht?

DerAndereAndi replied:
Hab ich gefixt, tut aber nicht weh. Danke!

DerAndereAndi committed:
Fix to lowercase

DerAndereAndi committed:
Fix crash

Fixes concurrent map read and map write issue

DerAndereAndi committed:
Fix charging state logic

DerAndereAndi committed:
Merge branch 'master' into feature/elli

DerAndereAndi committed:
Only set phases if they are wrong

This fixes the interval setting being ignored and everything going crazy

andig commented on charger/eebus.go:
```diff
@@ -125,6 +142,10 @@ func (c *EEBus) setLoadpointMinMaxLimits(data *communication.EVSEClientDataType)
+		return
+	}
+
 	if err := c.lp.SetPhases(int(data.EVData.ConnectedPhases)); err != nil {
```
Das ist unglücklich weil hier der Ladepunkt- so er das kann- physisch umgeschaltet wird was eigentlich völlig unnötig ist. Für's Erste sicher kein Problem, wäre aber gut da ein TODO ran zu hängen damit wir dynamische Phasen später mal von Auto nicht nur beim CONNECT sondern ständig abfragen.

DerAndereAndi replied:
Der Charger kann über EEBUS melden mit wie vielen Phasen er angeschlossen ist. Abklemmen, anderen Adapter dran, neu gestartet und Phasen sind physisch umgeschaltet. Hier geht es nicht um Umschaltungen während der Laufzeit. Das gibt EEBUS nicht her.

andig replied:
Das sind also nicht die Phasen des Fahrzeugs? Was passiert wenn nach 1p auf ein dummes 3p Fahrzeug gewechselt wird?

DerAndereAndi replied:
Das ist nur unterhalb von `EVData`, weil nur beim angeschlossenen EV die Daten übermittelt werden. Es handelt sich hier rein um die physische elektrische Anbindung der Box. Bei "normalen" Wallboxen ändert sich da nie was.

>  Was passiert wenn nach 1p auf ein dummes 3p Fahrzeug gewechselt wird?

Den Fall gibt es zur Laufzeit nicht.

andig replied:
Ich kann an der Wallbox kein anderes Auto anschliessen? Der Ladepunkt wird permanent umgeschaltet- was passiert damit beim Fahrzeugwechsel (disconnect, connect, identify)?

DerAndereAndi replied:
Wir verstehen uns mal wieder komplett nicht. Ich habe den Code entfernt, da die Funktion wohl (so denke ich mir das jetzt) nicht aussagt mit wie vielen Phasen die Wallbox angeschlossen ist, sondern wieviele Phasen für das Laden freigegeben sein sollen. Die Doku ist für mich hier nicht klar genug.

andig commented on charger/eebus.go:
```diff
@@ -515,18 +536,16 @@ func (c *EEBus) CurrentPower() (float64, error) {
 	var power float64
 	var phase uint
 	for phase = 1; phase <= data.EVData.ConnectedPhases; phase++ {
```
Slice mit Start bei 1? In Go wäre 0 üblich.

    for phase := uint(0)

DerAndereAndi replied:
Hier wird eine Map genutzt, kein Slice

andig commented on charger/eebus.go:
```diff
@@ -515,18 +536,16 @@ func (c *EEBus) CurrentPower() (float64, error) {
-		if phasePower, ok := data.EVData.Measurements.Power[phase]; ok {
-			power += phasePower
+		if phasePower, ok := data.EVData.Measurements.Power.Load(phase); ok {
+			power += phasePower.(float64)
```
Das führt zu panic falls mal kein float kommen sollte. Unschön.

DerAndereAndi replied:
Habe sanity checks hinzugefügt

DerAndereAndi committed:
Update isCharging detection and added tests

DerAndereAndi committed:
Add some sanity checks

DerAndereAndi committed:
Minor style change

DerAndereAndi commented:
Mein Ziel ist hier dass es funktioniert. Die EEBUS Implementierung und das momentane "Interface" sind für mich ein dead end. Möchte hier nicht mehr Zeit als notwendig einsetzen.

DerAndereAndi committed:
Remove commented dead code

DerAndereAndi committed:
Minor code style fix

DerAndereAndi commented:
Der Elli Charger Connect hat kein eingebautes Messgerät. Man kann dies selbst nachrüsten, dafür gibt es in Foren eine Anleitung. Der Benefit ist dass gemessene Stromstärken pro Phase gemeldet werden. Die Wallbox selbst liefert ansonsten einfach die vorgegebene Stromstärke zurück. Momentan wird über das Template das jedoch deaktiviert und der Anwender sieht keine Ladeleistung. Er könnte "einfach" das andere Template nehmen und würde dann Werte sehen.

Die Wallboxen sind alle installiert und ob der Besitzer jetzt noch einen Modbus Zähler mit mehr Daten einsetzt um dann auch Leistung und geladene Energiemenge zu erhalten, ist wohl nicht sehr wahrscheinlich. Die meisten Anwender sind keine DIY und haben wohl mit der Einrichtung von evcc schon genug Schwierigkeiten vor sich.

Meine Frage wäre nun: sollte man das im Template der Connect nicht einfach auch "erlauben", dass die theoretischen Werte geliefert werden?

@andig @premultiply was meint ihr?

andig commented:
Auf keinen Fall falls sie das nicht phasenrichtig macht.

DerAndereAndi commented:
Was meinst du damit? Es werden pro Phase die Stromstärke zurückgegeben, die als Limit gesetzt ist. Es ist ein theoretischer Wert.  Wenn das nicht erwünscht ist, bleibt es dann so wie es ist. Aber wie gesagt, der Anwender kann einfach das andere Template wählen und dann passiert genau das beschriebene. Es ist nicht möglich festzustellen ob da ein richtiger Zähler eingebaut ist oder nicht.

andig commented:
Wir verwenden die Ströme zur Phasenerkennung für den PV Modus. Die müssen stimmen, wenns nur L1 gibt darf auch nur auf L1 was kommen sonst funktioniert die Erkennung nicht.

premultiply commented:
Genau wie @andig es beschrieben hat. Auch wenn es auf den ersten Blick so aussieht wie dass was evcc intern dann auch macht.
Aber der entscheidende Unterschied ist, dass evcc in dem Fall dann glaubt, dass die übermittelten Phasenströme echt sind und damit die wichtige Phasenstromerkennung getäuscht wird und dann die ganze Regelung usw. nicht mehr richtig funktioniert.

DerAndereAndi committed:
Remove set phases

andig commented on charger/eebus.go:
```diff
@@ -42,18 +42,26 @@ func init() {
+		Meter         bool
+		ChargedEnergy bool
+	}{
+		Meter:         true,
```
Hier kein Default, dann kann brauchst Du im Template auch kein explizites `false` und das `eebusmeter` in der `default.yaml`.

DerAndereAndi committed:
Simplify templates

DerAndereAndi committed:
Update eebus and zeroconf

DerAndereAndi committed:
Support external connected meters

DerAndereAndi committed:
Add meter hint to Connect template

DerAndereAndi committed:
Fix for built in meters for isCharging check

DerAndereAndi committed:
Merge branch 'master' into feature/elli

DerAndereAndi committed:
Fix a crash

DerAndereAndi committed:
Fix detection of charging status with ext meter

The EV might charge with fewer phases than the ones the charger is connected with. If EEBUS doesn’t have that value, it assumes 3 phase connection, which might be wrong.

DerAndereAndi committed:
Update eebus lib

DerAndereAndi committed:
Merge branch 'master' into feature/elli

DerAndereAndi committed:
Change some errors to warnings

DerAndereAndi committed:
Change some warnings to trace

DerAndereAndi committed:
Fix comment

DerAndereAndi committed:
Merge branch 'master' into feature/elli

DerAndereAndi committed:
Add patching of asn1

This adds a `patch-asn1` task to the Makefile to patch `asn1.go` to allow the buggy Elli certificate to be accepted. The makefile task is invoked automatically on Github builds only

StefanSchoof commented on Makefile:
```diff
@@ -124,3 +130,9 @@ image-update:
+# patch asn1.go to allow Elli buggy certificates to be accepted with EEBUS
+patch-asn1:
+ifeq ($(shell test -e $(ASN1_FILE) && echo 1),1)
+	cd $(ASN1_PATH); sudo patch -N -t -i $(CURRENT_DIR)/patch/asn1.diff; cd $(CURRENT_DIR)
```
können wir die `--directory` Option von patch nutzen:
>        -d dir  or  --directory=dir
>          Change to the directory dir immediately, before doing anything
>          else.

DerAndereAndi committed:
Use the dir option of patch

DerAndereAndi committed:
Fix for previous commit

DerAndereAndi committed:
Minor template update

DerAndereAndi committed:
Merge branch 'master' into feature/elli

DerAndereAndi committed:
Fix self consumption support action

Only resend self consumption value, if the support value actually did change

DerAndereAndi committed:
Update eebus lib

DerAndereAndi committed:
Merge branch 'master' into feature/elli

DerAndereAndi committed:
Only check external meters every 10 seconds

DerAndereAndi committed:
Fix for isCharging check cache

DerAndereAndi committed:
Fix 10 minutes timeout workaround

DerAndereAndi committed:
Update eebus lib

DerAndereAndi committed:
Merge branch 'master' into feature/elli

DerAndereAndi committed:
Tidy mod

DerAndereAndi committed:
Merge branch 'master' into feature/elli

DerAndereAndi committed:
go mod tidy

DerAndereAndi committed:
Use evcc internal machineID when possible

DerAndereAndi committed:
Merge branch 'master' into feature/elli

andig commented on Makefile:
```diff
@@ -129,3 +135,9 @@ image-update:
+# patch asn1.go to allow Elli buggy certificates to be accepted with EEBUS
+patch-asn1:
+ifeq ($(shell test -e $(ASN1_FILE) && echo 1),1)
+	sudo patch -N -t -d $(ASN1_PATH) -i $(CURRENT_DIR)/patch/asn1.diff
```
Move sudo to calling location?

andig replied:
Dockerfile fehlt noch?

andig replied:
Können wir den Patch-PR vllt separieren und separat testen?

DerAndereAndi replied:
Na ja, dann hast mit dem PR hier die templates von Elli drin und werden in configure angezeigt, aber die nightly oder release builds unterstützen das nicht.

Ich würde es daher präferieren das hier alles zusammen abzuhandeln

DerAndereAndi replied:
What exactly do you mean by this: `Move sudo to calling location?` ?

premultiply replied:
Ich würde vermuten: Das `sudo` hier an dieser Stelle weg und das ganze Makefile ggf. mit sudo (oder anderen ausreichenden Rechten) ausführen.
```suggestion
	patch -N -t -d $(ASN1_PATH) -i $(CURRENT_DIR)/patch/asn1.diff
```

DerAndereAndi replied:
Hmm .... In Zeile 125 wird es ebenfalls schon genau so verwendet. Den make bin sudo aufrufen ... fände ich jetzt merkwürdig

StefanSchoof replied:
Da in docker build normal kein `sudo` vorhanden ist, würde der Aufruf in `Dockerfile` nicht funktionieren.

DerAndereAndi replied:
Ok, also Aufruf mit `sudo` in den CI workflows only

DerAndereAndi replied:
Ich habe das mal hier geändert: https://github.com/evcc-io/evcc/pull/4207/commits/e7a609b74782e3d832ae334494ba08ed1303e467

Und das patchen noch zum `Build` job hinzugefügt, aber ich vermute mal das ist unnötig weil der `clean` job ja immer davor ausgeführt werden sollte

andig commented on charger/eebus.go:
```diff
@@ -143,34 +160,38 @@ func (c *EEBus) showCurrentChargingSetup() {
 		c.communicationStandard = data.EVData.CommunicationStandard
 		timestamp := time.Now()
-		c.log.WARN.Println("!! ", timestamp.Format("2006-01-02 15:04:05"), " ev-charger-communication changed from ", prevComStandard, " to ", data.EVData.CommunicationStandard)
+		c.log.TRACE.Println("!! ", timestamp.Format("2006-01-02 15:04:05"), " ev-charger-communication changed from ", prevComStandard, " to ", data.EVData.CommunicationStandard)
```
brauchen wir die Timestamps jeweils doppelt zum Timestamp im Log?

DerAndereAndi replied:
Beim Warning war es sinnvoll da es in der UI angezeigt wurde, jetzt nicht mehr. Wird entfernt

andig commented on charger/eebus.go:
```diff
@@ -143,34 +160,38 @@ func (c *EEBus) showCurrentChargingSetup() {
-			c.log.ERROR.Println("failed to send current limit data: ", err)
+		// but only if the support value actually changed
+		if prevSelfConsumptionSupport != c.selfConsumptionSupportAvailable {
+			err := c.writeCurrentLimitData([]float64{c.maxCurrent, c.maxCurrent, c.maxCurrent})
```
if err := ...; err !=nil {

andig commented on go.mod:
```diff
@@ -173,3 +174,5 @@ require (
 
 replace github.com/foogod/go-powerwall => github.com/andig/go-powerwall v0.2.1-0.20220205120646-e5220ad9a9a0
+
+replace github.com/evcc-io/eebus => github.com/DerAndereAndi/eebus v0.0.0-20220928183324-ef68aab01342
```
Können wir den vorab mergen?

DerAndereAndi replied:
Das sollte gemacht werden. Deshalb ist der PR dort nicht mehr im Draft modus aber dieser hier schon

andig replied:
Kann ich den EEBus PR schon mergen ohne dass der evcc Build kaputt geht? Dann wär das der nächste Schritt?

DerAndereAndi replied:
Klar, es wird ja in master ein älterer Commit referenziert

andig replied:
Stimm auch wieder. Danke :)

DerAndereAndi committed:
Minor style fixes

DerAndereAndi committed:
Update patching go

- Remove sudo from the makefile
- Call `sudo make patch-asn1` from the default workflow build job

andig commented:
Bzgl. ASN: jetzt fehlt noch Docker?

DerAndereAndi commented:
Wieso? Der läuft als Teil des GitHub Workflows, und da werden die Jobs in `default.yaml` durchgeführt in der go installiert und dann gepatched wird.

andig commented:
> Wieso? Der läuft als Teil des GitHub Workflows, und da werden die Jobs in `default.yaml` durchgeführt in der go installiert und dann gepatched wird.

Nein. Der Docker Build läuft in einem Docker Container.

andig commented:
Nämlich hier: https://github.com/evcc-io/evcc/blob/master/Dockerfile#L23

DerAndereAndi commented:
Ok, was genau soll da jetzt wo rein? Ich kenne diesen Teil nicht.

andig committed:
Add docker patch

andig commented:
> Ok, was genau soll da jetzt wo rein?

Done

DerAndereAndi committed:
Update eebus lib

DerAndereAndi committed:
Remove old UniqueID usage

DerAndereAndi committed:
Minor error handling improvement

DerAndereAndi committed:
Update eebus lib to master

DerAndereAndi committed:
Update Elli templates marking them as beta

DerAndereAndi committed:
Remove more additional log timestamps

DerAndereAndi committed:
Update eebus library

andig commented on Makefile:
```diff
@@ -129,3 +135,9 @@ image-update:
+
+# patch asn1.go to allow Elli buggy certificates to be accepted with EEBUS
+patch-asn1:
+ifeq ($(shell test -e $(ASN1_FILE) && echo 1),1)
```
Ist das nicht eigentlich:

    shell test -e $(ASN1_FILE) && patch -N -t -d $(ASN1_PATH) -i $(CURRENT_DIR)/patch/asn1.diff

DerAndereAndi replied:
Stimmt :) Manchmal sieht man den Wald vor lauter Bäumen nicht

andig committed:
Simplify patch and ignore already-applied error
